File: elixir/aoc/lib/aoc/day04a.ex
defmodule AOC.Day04a do
  use AOC

  def solution(path) do
    path
    |> read_lines()
    |> Enum.map(&String.trim/1)
    |> Enum.map(&String.split(&1, ": "))
    |> Enum.map(fn [_card, numbers] -> numbers end)
    |> Enum.map(&String.split(&1, " | "))
    |> Enum.map(&score_card/1)
    |> Enum.sum()
  end

  def score_card([winning, ticket]) do
    winning_numbers = winning |> String.split(" ", trim: true)
    ticket_numbers = ticket |> String.split(" ", trim: true)

    ticket_numbers
    |> Enum.filter(&Enum.member?(winning_numbers, &1))
    |> Enum.count()
    |> matches_to_score()
  end

  def matches_to_score(0), do: 0
  def matches_to_score(1), do: 1
  def matches_to_score(n), do: 2 ** (n - 1)
end

File: ruby/aoc/lib/aoc/day04a.rb
module AOC
  class Day04a < Day
    class Card
      attr_reader :id
      attr_reader :winning_numbers
      attr_reader :selected_numbers

      def initialize(raw)
        game_ref, numbers = raw.split(": ")
        @id = game_ref.split(" ").last.to_i
        winning_numbers, selected_numbers = numbers.split(" | ")
        @winning_numbers = winning_numbers.split(" ").map(&:to_i)
        @selected_numbers = selected_numbers.split(" ").map(&:to_i)
      end

      def matches
        @matches ||= (@winning_numbers & @selected_numbers).length
      end

      def score
        return 0 if matches == 0
        return 1 if matches == 1
        return 2 ** (matches - 1)
      end
    end

    def sum_scores
      cards.map(&:score).sum
    end

    def cards
      @cards ||= File.readlines(@path, chomp: true).map do |line|
        Card.new(line)
      end
    end
  end
end
